f90wrap_helpers/classes.py

```diff
@@ -0,0 +1,84 @@
+from pyhmcode.hmx import HMcode2016 as HMcode2016
+from pyhmcode.hmx import HMcode2020 as HMcode2020
+from pyhmcode.hmx import HMcode2020_feedback as HMcode2020_feedback
+
+from pyhmcode.hmx import HMx2020_matter_pressure_w_temp_scaling as HMx2020_matter_pressure_w_temp_scaling
+from pyhmcode.hmx import HMx2020_matter_w_temp_scaling as HMx2020_matter_w_temp_scaling
+
+from pyhmcode.hmx import field_dmonly as field_dmonly
+from pyhmcode.hmx import field_matter as field_matter
+from pyhmcode.hmx import field_cdm as field_cdm
+from pyhmcode.hmx import field_gas as field_gas
+from pyhmcode.hmx import field_stars as field_stars
+from pyhmcode.hmx import field_electron_pressure as field_electron_pressure
+
+class Cosmology(pyhmcode.cosmology_functions.cosmology):
+    def __init__(self):
+        super().__init__()
+
+        # Assign reasonable default parameters (icosmo=1 == "boring cosmology")
+        pyhmcode.cosmology_functions.assign_cosmology(icosmo=1, cosm=self, verbose=False)
+
+        self.norm_method = pyhmcode.cosmology_functions.norm_none
+        self.itk = pyhmcode.cosmology_functions.iTk_external
+        self.has_power = True
+
+        self._has_linear_power_set = False
+
+    def set_linear_power_spectrum(self, k, z, pofk_lin):
+        if pofk_lin.shape != (len(z), len(k)):
+            raise ValueError(f"Shape mismatch between k, z, pofk_lin: {k.shape}, {z.shape}, {pofk_lin.shape}")
+        if z[0] >= z[-1]:
+            raise ValueError("Redshifts need to be in increasing order.")
+
+        pk_lin = np.asfortranarray(pofk_lin[::-1].T, dtype=np.float64)
+        self.a_lin = np.asfortranarray((1/(1+z))[::-1], dtype=np.float64)
+        self.k_lin = np.asfortranarray(k, dtype=np.float64)
+
+        pyhmcode.cosmology_functions.init_external_linear_power_tables(self, self.k_lin, self.a_lin, pk_lin)
+        pyhmcode.cosmology_functions.init_cosmology(self)
+
+        self._has_linear_power_set = True
+
+class Halomodel(pyhmcode.hmx.halomod):
+    def __new__(cls, mode, verbose=False):
+        # Need to use __new__ to intercept creation of the instance and use
+        # assign_halomod instead
+        instance = pyhmcode.hmx.assign_halomod(mode, verbose)
+        return instance
+
+
+
+def calculate_nonlinear_power_spectrum(cosmology, halomodel, fields=np.array([field_dmonly]),
+                                       return_halo_terms=False, verbose=False):
+    if not cosmology._has_linear_power_set:
+        raise RuntimeError("Cosmology has no linear power spectrum set.")
+
+    nk = len(cosmology.k_lin)
+    na = len(cosmology.a_lin)
+    nf = len(fields)
+
+    # Output arrays
+    pow_lin = np.zeros((nk, na), dtype=np.float64, order="F")
+    pow_2h = np.zeros((nf, nf, nk, na), dtype=np.float64, order="F")
+    pow_1h = np.zeros((nf, nf, nk, na), dtype=np.float64, order="F")
+    pow_hm = np.zeros((nf, nf, nk, na), dtype=np.float64, order="F")
+
+    pyhmcode.hmx.calculate_hmx_old(ifield=fields, nf=nf, 
```

Comment:
Do you need to use `calculate_hmx_old` here? Or would it be possible to upgrade to one of the newer routines? If this is going to be a python portal into pure `HMcode` then I don't think we need to worry about the `fields` types, because this is always set to `DMONLY` for `HMcode`. But if we also want to expose the `HMx` functionality then I guess we need to worry about the fields.

Reply:
Yes, unless there's another function that serves the same purpose (i.e., not calling assign_cosmology and assign_halomod inside). 
The idea is to have unified interface to HMCode and HMx, with defaulting to HMCode if no fields are specified (this might not be in the PR yet). For people that only care about a vanilla matter-matter P(k), using the CAMB version is probably easier. 
I see `pyhmcode` really for people that need advanced functionality.

Reply:
Ah, I see. And you use `calculate_hmx_old` because you assign the halo model somewhere else? And this is in case you then want to change the properties of the halo models somewhere? I was going to suggest that you change to `calculate_HMx_full`, but  this assigns the halo model internally and so the user couldn't change any of the properties.

Reply:
Exactly. If someone wants to just run with the default parameters, something like the CAMB implementation is more appropriate.